server/src/test/java/org/opensearch/search/pipeline/SearchPipelineStatsTests.java

```diff
@@ -0,0 +1,183 @@
+import org.opensearch.test.OpenSearchTestCase;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+
+public class SearchPipelineStatsTests extends OpenSearchTestCase {
+    public void testSerializationRoundtrip() throws IOException {
+        SearchPipelineStats stats = createStats();
+        SearchPipelineStats deserialized;
+        try (BytesStreamOutput bytesStreamOutput = new BytesStreamOutput()) {
+            stats.writeTo(bytesStreamOutput);
+            try (StreamInput bytesStreamInput = bytesStreamOutput.bytes().streamInput()) {
+                deserialized = new SearchPipelineStats(bytesStreamInput);
+            }
+        }
+        assertEquals(stats, deserialized);
+    }
+
+    private static SearchPipelineStats createStats() {
+        return new SearchPipelineStats(
+            new SearchPipelineStats.Stats(1, 2, 3, 4),
+            new SearchPipelineStats.Stats(5, 6, 7, 8),
+            List.of(
+                new SearchPipelineStats.PipelineStats(
+                    "p1",
+                    new SearchPipelineStats.Stats(9, 10, 11, 12),
+                    new SearchPipelineStats.Stats(13, 14, 15, 16)
+                ),
+                new SearchPipelineStats.PipelineStats(
+                    "p2",
+                    new SearchPipelineStats.Stats(17, 18, 19, 20),
+                    new SearchPipelineStats.Stats(21, 22, 23, 24)
+                )
+
+            ),
+            Map.of(
+                "p1",
+                new SearchPipelineStats.PipelineDetailStats(
+                    List.of(new SearchPipelineStats.ProcessorStats("req1:a", "req1", new SearchPipelineStats.Stats(25, 26, 27, 28))),
+                    List.of(new SearchPipelineStats.ProcessorStats("rsp1:a", "rsp1", new SearchPipelineStats.Stats(29, 30, 31, 32)))
+                ),
+                "p2",
+                new SearchPipelineStats.PipelineDetailStats(
+                    List.of(
+                        new SearchPipelineStats.ProcessorStats("req1:a", "req1", new SearchPipelineStats.Stats(33, 34, 35, 36)),
+                        new SearchPipelineStats.ProcessorStats("req2", "req2", new SearchPipelineStats.Stats(37, 38, 39, 40))
+                    ),
+                    List.of()
+                )
+            )
+        );
+    }
+
+    public void testToXContent() throws IOException {
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        XContentBuilder xContentBuilder = new XContentBuilder(JsonXContent.jsonXContent, bos);
+        xContentBuilder.prettyPrint();
+        xContentBuilder.startObject();
+        createStats().toXContent(xContentBuilder, null);
+        xContentBuilder.endObject();
+        xContentBuilder.close();
+
+        String jsonContent = bos.toString(StandardCharsets.UTF_8);
+
+        assertEquals(
+            "{\n"
```

Comment:
Isn't there a better way of testing JSON content? To me this seems to be too sensitive to formatting.

Reply:
That's a good point -- depending on the JSON formatting is risky and brittle.

My concerns with recreating the structure with XContentBuilder methods are:
1. To some extent, we're duplicating the logic that we're testing (without the conditionals/loops).
2. The JSON is "arguably" more readable -- I used this test as a confirmation that the response looks how I expected it. (I did want to use a `"""` string block to hide the  escaped quotes + newlines, but I think we're still on language level 11.)

What I managed to do to avoid depending on the exact JSON string is:

```
String expected = "...giant JSON string...";
XContentParser expectedParser = JsonXContent.jsonXContent.createParser(this.xContentRegistry(), DeprecationHandler.THROW_UNSUPPORTED_OPERATION, expected);
XContentBuilder expectedBuilder = XContentBuilder.builder(JsonXContent.jsonXContent);
expectedBuilder.generator().copyCurrentStructure(expectedParser);

assertEquals(XContentHelper.convertToMap(BytesReference.bytes(expectedBuilder), false, (MediaType) XContentType.JSON),
             XContentHelper.convertToMap(BytesReference.bytes(actualBuilder), false, (MediaType) XContentType.JSON));
```

It removes the format sensitivity but still serves as a "preview" of what the JSON looks like. What do you think?